lib/src/value/value.ts

```diff
@@ -0,0 +1,151 @@
+// Copyright 2021 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import {List, OrderedMap, ValueObject} from 'immutable';
+import {valueError} from '../utils';
+
+/**
+ * A SassScript value.
+ *
+ * All SassScript values are immutable.
+ *
+ * Concrete values (such as `SassColor`) are implemented as subclasses and get
+ * instantiated as normal JS classes.
+ *
+ * Untyped values can be cast to particular types using `assert*()` functions,
+ * which throw user-friendly error messages if they fail.
+ *
+ * All values, except `false` and `null`, count as `true`.
+ *
+ * All values can be used as lists. Maps count as lists of pairs, while all
+ * other values count as single-value lists.
+ */
+export abstract class Value implements ValueObject {
+  /** Whether `this` counts as `true`. */
+  get isTruthy(): boolean {
+    return true;
+  }
+
+  /** Returns JS null if `this` is `sassNull`. Otherwise, returns `this`. */
+  get realNull(): Value | null {
+    return this;
+  }
+
+  private _asList?: List<Value>;
+
+  /** `this` as a list. */
+  get asList(): List<Value> {
+    return this._asList ?? (this._asList = List([this]));
+  }
+
+  /** The separator for `this` as a list. */
+  get separator(): null {
+    return null;
+    // TODO(awjin): Implement the proper return type ListSeparator.
+  }
+
+  /** Whether `this`, as a list, has brackets. */
+  get hasBrackets(): boolean {
+    return false;
+  }
+
+  /**
+   * Converts `sassIndex` to a JS index into the array returned by `asList`.
+   *
+   * Sass indices start counting at 1, and may be negative in order to index
+   * from the end of the list.
+   *
+   * `sassIndex` must be...
+   * - a number, and
+   * - an integer, and
+   * - a valid index into `asList`.
+   *
+   * Otherwise, this throws an error.
+   */
+  sassIndexToListIndex(sassIndex: Value): number {
```

Comment:
Since this verifies details about `sassIndex`, it should take a `name` parameter. Can you update the spec?

Reply:
Yep, PR here: https://github.com/sass/sass/pull/3111